File: core/core/ImageViews.m
////
////  CacheImage.swift
////  LinCore
////
////  Created by lin on 1/18/15.
////  Copyright (c) 2015 lin. All rights reserved.
////
//
//
//#import <UIKit/UIKit.h>
//#import "ImageViews.h"
////#import <LinUtil/LinUtil.h>
////#import "Lin
//
//
//@interface CacheImageOperation : NSOperation
//
//@end
//
//@interface CacheImageOperation(){
//    @private
//    NSString * _path;
//    CacheImageView * _imageView;;
//}
//
//@end
//
//
//@interface CacheImageView (){
//@private
//    BOOL isImage;
//    void (^_imageChanged)(CacheImageView * imageView);
//    NSRecursiveLock * lock;
//    CacheImageOperation * operation;
//    NSString * _path;
//}
//
//-(void)setCacheImage:(UIImage *)image;
//@end
//
//
//
//
//@implementation CacheImageOperation
//
//
//-(id)initWithPath:(NSString*)path imageView:(CacheImageView*)imageView{
//    self = [super init];
//    
//    if (self) {
//        self->_path = path;
//        self->_imageView = imageView;
//    }
//    
//    return self;
//}
//
//-(void)main{
//    BOOL isCancelled = [self isCancelled];
//        if (isCancelled) {
//            return;
//        }
//
//    __block NSData * data = nil;
//    NSURL * url = [[NSURL alloc] initWithString:self->_path];
//    if (url != nil) {
//        data = [CacheImageView cachedataForUrl:url];
//    }
//    
//    
//    dispatch_async(dispatch_get_main_queue(), ^{
//            if (![self isCancelled]) {
//                if (data != nil) {
//                    [self->_imageView setCacheImage:[[UIImage alloc] initWithData:data]];
//                }else{
//                    if (_imageView.noImage) {
//                        [self->_imageView setCacheImage:[UIImage imageNamed:_imageView.noImage]];
//                    }else{
//                        [self->_imageView setCacheImage:nil];
//                    }
//                }
//            }
//        });
//
//}
//
//@end
//NSString * CacheImageView_Cache_Path_tmpvar;
//
//
//dispatch_once_t ___CacheImageView_predicate = 0;
//Queue * ___CacheImageView_instance;
//
//@implementation CacheImageView
//
//
//-(void (^)(CacheImageView * imageView))imageChanged{
//    return self->_imageChanged;
//}
//
//-(void)setImageChanged:(void (^)(CacheImageView * imageView))imageChanged{
//    self->_imageChanged = imageChanged;
//}
//
//+(NSData *)cachedataForUrl:(NSURL *)url{
//    if (url == nil) {
//        return nil;
//    }
//    NSString * md5 = [url absoluteString].md5;
//    NSString * filename;
//    if ([CacheImageView.cachePath hasSuffix:@"/"]) {
//        filename = [[NSString alloc] initWithFormat:@"%@%@", CacheImageView.cachePath , md5 ];
//    }else{
//        filename = [[NSString alloc] initWithFormat:@"%@/%@", CacheImageView.cachePath , md5 ];
//    }
//
//    NSData * data = [NSData dataWithContentsOfFile:filename];
//    if (data == nil) {
//        data = [NSData dataWithContentsOfURL:url];
//        [data writeToFile:filename atomically:TRUE];
//    }
//    return data;
//
//}
//
//+(NSString *)cachePath{
//    if (CacheImageView_Cache_Path_tmpvar == nil) {
//        CacheImageView_Cache_Path_tmpvar = pathFor(DocumentsCache, @"imagecache");
//    }
//    [self createCachePath];
//    return CacheImageView_Cache_Path_tmpvar;
//}
//
//+(void)setCachePath:(NSString *)cachePath{
//    CacheImageView_Cache_Path_tmpvar = cachePath;
//    [self createCachePath];
//}
//
//+(void)createCachePath{
//    NSFileManager * fileManager = [NSFileManager defaultManager];
//    BOOL isDir = false;
//    if (![fileManager fileExistsAtPath:CacheImageView_Cache_Path_tmpvar isDirectory:&isDir] || isDir == false) {
//        [fileManager createDirectoryAtPath:CacheImageView_Cache_Path_tmpvar withIntermediateDirectories:TRUE attributes:nil error:nil];
//    }
//}
//
//
//+(Queue*)imageCacheQueue{
//    dispatch_once(&___CacheImageView_predicate,^{
//        ___CacheImageView_instance = [[Queue alloc] initWithCount:20];
//    });
//    return ___CacheImageView_instance;
//}
//
//-(NSString *)path{
//    return self->_path;
//}
//-(void)setPath:(NSString *)path{
//    if (path == self->_path && self->isImage == false) {
//        self->_path = path;
//        return;
//    }
//    self->_path = path;
//    [self loadImage];
//}
//-(id)initWithPath:(NSString*)path{
//    self = [super init];
//    if (self) {
//        self->lock = [[NSRecursiveLock alloc] init];
//        self->isImage = false;
//        self->_path = path;
//    }
//    return self;
//}
//-(id)init{
////    self = [super initWithFrame:CGRectMake(0, 0, 0, 0)];
//    self = [super init];
//    if (self) {
//        self->lock = [[NSRecursiveLock alloc] init];
//        self->isImage = false;
//    }
//    return self;
//}
//
//-(id)initWithFrame:(CGRect)frame{
//    self = [super initWithFrame:frame];
//    if (self) {
//        self->lock = [[NSRecursiveLock alloc] init];
//        self->isImage = false;
//    }
//    return self;
//}
//
//-(void)loadImage{
//    [self->lock lock];
//    
//    [self->operation cancel];
//    super.image = nil;
//    if (self.path != nil && ![self.path isKindOfClass:[NSNull class]]) {
//        isImage = false;
//        if ([self.path hasPrefix:@"http://"]) {
//            operation = [[CacheImageOperation alloc] initWithPath:self.path imageView:self];
//            [CacheImageView.imageCacheQueue addOperation:operation];
//        }else{
//            [self setImage:[UIImage imageNamed:self.path]];
//        }
//        
//    }else{
//        [self setImage:[UIImage imageNamed:self.noImage]];
//    }
//    [self->lock unlock];
//}
//
//-(UIImage *)image{
//    return super.image;
//}
//
//-(void)setImage:(UIImage *)image{
//    [self->lock lock];
//    isImage = true;
//    [self->operation cancel];
//    super.image = image;
//    [lock unlock];
//    if (_imageChanged != nil) {
//        _imageChanged(self);
//    }
//}
//
//-(void)setCacheImage:(UIImage *)image{
//    [lock lock];
//    if (isImage == false) {
//        super.image = image;
//    }
//    [lock unlock];
//    if (_imageChanged != nil) {
//        _imageChanged(self);
//    }
//}
//
//-(void)removeFromSuperview{
//    [operation cancel];
//}
//
//@end
//
//
//
//@interface FillImageView (){
//    @private
//    CGSize preImageSize;
//    CGSize preVieSize;
//    CacheImageView * _imageView;
//    NSObject * _image;
//    ImageFill _fill;
//    
//    void (^_imageChanged)();
//}
//@end
//
//@implementation FillImageView
//
//
//-(UIImageView *)imageView{
//    return _imageView;
//}
//
//-(void (^)())imageChanged{
//    return _imageChanged;
//}
//
//-(void)setImageChanged:(void (^)())imageChanged{
//    _imageChanged = imageChanged;
//}
//
//-(ImageFill)fill{
//    return _fill;
//}
//
//-(void)setFill:(ImageFill)fill{
//    if (_fill == fill) {
//        return;
//    }
//    _fill = fill;
//    [self update];
//}
//
//-(instancetype)init{
//    self = [super init];
//    
//    if (self) {
//        _imageView = [[CacheImageView alloc] init];
//        
//        __weak FillImageView * wself = self;
//        _imageView.imageChanged = ^(CacheImageView * _){
//            FillImageView * sself = wself;
//            
//            [sself update];
//            
//            if (sself->_imageChanged != nil) {
//                sself->_imageChanged();
//            }
//        };
//        
//        [self addSubview:_imageView];
//        self.backgroundColor = [UIColor clearColor];
//    }
//    
//    return self;
//}
//
//-(NSObject *)image{
//    return _image;
//}
//
//-(void)setImage:(NSObject *)image{
//    if (_image == image) {
//        return;
//    }
//    _image = image;
//    if (image == nil) {
//        _imageView.image = nil;
//        return;
//    }
//    if ([image isKindOfClass:[NSString class]]) {
//        _imageView.path = (NSString*)image;
//    }else if([image isKindOfClass:[UIImage class]]){
//        _imageView.image = (UIImage*)image;
//    }else{
//        _imageView.image = nil;
//    }
//}
//
//-(void)update{
//
//    if (_imageView.image == nil) {
//        return;
//    }
//    
//    CGRect viewRect = self.bounds;
//    
//    CGSize imageSize = _imageView.image.size;
//    if (imageSize.width == 0 || imageSize.height == 0) {
//        return;
//    }
//    
//    if (viewRect.size.width == 0 || viewRect.size.height == 0) {
//        viewRect = CGRectMake(0, 0, imageSize.width, imageSize.height);
//        [self setFrame:viewRect];
////        [self layoutIfNeeded];
//    }
//    CGRect contentViewRect = viewRect;
//    if (imageSize.width != preImageSize.width || imageSize.height != preImageSize.height
//        || contentViewRect.size.width != preVieSize.width || contentViewRect.size.height != preVieSize.height) {
//        preImageSize = imageSize;
//        preVieSize = contentViewRect.size;
//        CGFloat s = 1.0;
//        if (self.fill != ImageFillFill) {
//            s = (preImageSize.width/preImageSize.height)/(contentViewRect.size.width/contentViewRect.size.height);
//        }
//        if ((ImageFillDefault == self.fill && s > 1) || self.fill == ImageFillWithWidth) {
//            contentViewRect.size.width = contentViewRect.size.width;
//            contentViewRect.size.height = contentViewRect.size.height / s;
//            contentViewRect.origin.x = 0;
//            contentViewRect.origin.y = (viewRect.size.height - contentViewRect.size.height) / 2;
//        }else{
//            contentViewRect.size.width = contentViewRect.size.width * s;
//            contentViewRect.size.height = contentViewRect.size.height;
//            contentViewRect.origin.y = 0;
//            contentViewRect.origin.x = (viewRect.size.width - contentViewRect.size.width) / 2;
//        }
//        _imageView.frame = contentViewRect;
//    
//    }
//}
//
//-(void)layoutSubviews{
//    [self update];
//    [super layoutSubviews];
//}
//
//@end
//

File: core/core/ImageViews.swift
//
//  CacheImage.swift
//  LinCore
//
//  Created by lin on 1/18/15.
//  Copyright (c) 2015 lin. All rights reserved.
//


import UIKit
import LinUtil

public enum ImageFill: Int {
    case `default` = 0
    case fill = 1
    case width = 2
    case height = 4
}

//typedef NS_ENUM(NSUInteger, ImageFill) {
//    ImageFillDefault,    //
//    ImageFillFill,  //
//    ImageFillWithWidth,
//    ImageFillWithHeight//
//} ;
//@interface CacheImageOperation : NSOperation
//
//@end
//
//@interface CacheImageOperation(){
//    @private
//    NSString * _path;
//    CacheImageView * _imageView;;
//}
//
//@end


//@interface CacheImageView (){
//@private
//    BOOL isImage;
//    void (^_imageChanged)(CacheImageView * imageView);
//    NSRecursiveLock * lock;
//    CacheImageOperation * operation;
//    NSString * _path;
//}
//
//-(void)setCacheImage:(UIImage *)image;
//@end




//public class CacheImageOperation : NSOperation{
//    
//    private var _path:String;
//    private var _imageView:CacheImageView;;
//
//
//    public init(path:String, imageView:CacheImageView){
////    self = [super init];
//    
//        self._path = path;
//        self._imageView = imageView;
//    }
//
//    public override func main(){
//        let isCancelled = super.cancelled
//        if isCancelled {
//            return;
//        }
//
//        var data:NSData? = nil;
//        let url = NSURL(string:self._path);
//        if let url = url {
//            data = CacheImageView.cachedataForUrl(url);
//        }
//    
//    
//    dispatch_async(dispatch_get_main_queue()) {[weak self]in
//            if self?.cancelled ?? false {
//                if data != nil {
//                    self?._imageView.image = UIImage(data:data!);
//                }else{
//                    if self?._imageView.noImage == nil ?? false {
//                        self._imageView.setCacheImage(UIImage(named:_imageView.noImage));
//                    }else{
//                        self._imageView.setCacheImage(nil);
//                    }
//                }
//            }
//        };
//
//}
//
//}

//private var _CacheImageView_Cache_Path_tmpvar:String?;
//
//
//private var ___CacheImageView_predicate:dispatch_once_t = 0;
//private var ___CacheImageView_instance:Queue!;
//
//
//public class CacheImageView : UIImageView{
//    
////}
////
////@implementation CacheImageView
//    private var _isImage:Bool = false;
//    //    void (^_imageChanged)(CacheImageView * imageView);
//    private var _imageChanged:((imaegView:CacheImageView)->())?
//    private var _lock:NSRecursiveLock?;
//    private var _operation:CacheImageOperation!;
////    private var _path:String?;
//
//
////-(void (^)(CacheImageView * imageView))imageChanged{
////    return self->_imageChanged;
////}
////
////-(void)setImageChanged:(void (^)(CacheImageView * imageView))imageChanged{
////    self->_imageChanged = imageChanged;
////}
//    private var imageChanged:((imaegView:CacheImageView)->())?;
//
//    private class func cachedataForUrl(url:NSURL?)->NSData?{
//    if (url == nil) {
//        return nil;
//    }
//        let md5 = url.absoluteString.md5;
//        var filename:String!;
//    if  CacheImageView.cachePath.hasSuffix("/") {
//        filename = "\(CacheImageView.cachePath)\(md5)";
//    }else{
//        filename = "\(CacheImageView.cachePath)/\(md5)";
//    }
//
//    let data = NSData(file:filename);
//    if data == nil {
//        data = NSData(URL:url);
//        data.writeToFile(filename, atomically:true);
//    }
//    return data;
//
//}
//
//    private class var cachePath:String{
//        get{
//    if _CacheImageView_Cache_Path_tmpvar == nil {
//        _CacheImageView_Cache_Path_tmpvar = pathFor(DocumentsCache, "imagecache");
//    }
//    self.createCachePath();
//    return _CacheImageView_Cache_Path_tmpvar;
//}
//
//set{
//    CacheImageView_Cache_Path_tmpvar = newValue;
//    self.createCachePath();
//}
//    }
//
//private class func createCachePath(){
//    let fileManager = NSFileManager.defaultManager();
////    BOOL isDir = false;
//    let isDir = UnsafeMutablePointer<ObjCBool>.alloc(1);
//    if !fileManager.fileExistsAtPath(_CacheImageView_Cache_Path_tmpvar, isDirectory:isDir) || isDir.memory == false {
//        fileManager.createDirectoryAtPath(_CacheImageView_Cache_Path_tmpvar, withIntermediateDirectories:true, attributes:nil, error:nil);
//    }
//}
//
//
//private func imageCacheQueue()->Queue{
//    dispatch_once(&___CacheImageView_predicate,^{
//        ___CacheImageView_instance = Queue(count:20);
//    });
//    return ___CacheImageView_instance;
//}
//
//    public var path:String? = nil{
////        get{
////    return self->_path;
////}
////set{
//        didSet{
//    if (oldValue == self.path && self._isImage == false) {
////        self->_path = newValue;
//        return;
//    }
////    self->_path = newValue;
//    self.loadImage();
//}
//    }
//
//        public init(path:String){
////    self = [super init];
////    if (self) {
//        self._lock = NSRecursiveLock();
//        self._isImage = false;
//        self.path = path;
////    }
////    return self;
//}
//    public init(){
////    self = [super initWithFrame:CGRectMake(0, 0, 0, 0)];
////    self = [super init];
////    if (self) {
//        self._lock = NSRecursiveLock();
//        self._isImage = false;
////    }
////    return self;
//}
//
//    public init(frame:CGRect){
////    if (self) {
//        self._lock = NSRecursiveLock();
//        self._isImage = false;
////    }
//    super.init(frame:frame);
////    return self;
//}
//
//private func loadImage(){
//    self._lock.lock();
//    
//    self._operation.cancel();
//    super.image = nil;
//    if self.path != nil && !self.path.isKindOfClass(NSNull.self) {
//        isImage = false;
//        if self.path!.hasPrefix("http://") {
//            operation = CacheImageOperation(path:self.path, imageView:self);
//            CacheImageView.imageCacheQueue.addOperation(operation);
//        }else{
//            self.setImage(UIImage(named:self.path));
//        }
//        
//    }else{
//        self.setImage(UIImage(named:self.noImage));
//    }
//    self._lock.unlock();
//}
//
//    public var image:UIImage?{
//        get{
//    return super.image;
//}
//
//set{
//    self._lock.lock();
//    isImage = true;
//    self._operation.cancel();
//    super.image = newValue;
//    lock.unlock();
//    if _imageChanged != nil {
//        _imageChanged(self);
//    }
//}
//    }
//
//    private func setCacheImage(iamge:UIImage){
//    lock.lock();
//    if _isImage == false {
//        super.image = image;
//    }
//    lock.unlock();
//    if _imageChanged != nil {
//        _imageChanged(self);
//    }
//}
//
//public func removeFromSuperview(){
//    _operation.cancel();
//}
//}



open class FillImageView : UIImageView{
    fileprivate var _preImageSize:CGSize = CGSize(width: 0, height: 0);
    fileprivate var _preVieSize:CGSize = CGSize(width: 0, height: 0);
    fileprivate var _imageView:CacheImageView;
//    private var _image:AnyObject;
//    private var _fill:ImageFill;
    
//    void (^_imageChanged)();
//}
//@end
    open var imageChanged:(()->())?

//@implementation FillImageView


    open var imageView:UIImageView{
    return _imageView;
}

//-(void (^)())imageChanged{
//    return _imageChanged;
//}
//
//-(void)setImageChanged:(void (^)())imageChanged{
//    _imageChanged = imageChanged;
//}

    open var fill:ImageFill = ImageFill.default{
        didSet{
    if (oldValue == fill) {
        return;
    }
    self.update();
}
    }

    public init(){
//    self = [super init];
//    
//    if (self) {
        _imageView = CacheImageView();
        
        super.init(frame: CGRect(x: 0, y: 0, width: 0, height: 0));
//        __weak FillImageView * wself = self;
        _imageView.imageChanged = {[weak self] (_:CacheImageView)->() in
//            FillImageView * sself = wself;
            
            self?.update();
            
            if let imageChanged = self?.imageChanged {
                imageChanged();
            }
        };
        
        self.addSubview(_imageView);
        self.backgroundColor = UIColor.clear;
    }
    
    required public init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
//    return self;
//}

//-(NSObject *)image{
//    return _image;
//}

    open func setFillImage(_ image:AnyObject?){
//        didSet{
    
//        }
//    }
//-(void)setImage:(NSObject *)image{
    if image == nil {
        return;
    }
//    if self.image == nil {
//        _imageView.image = nil;
//        return;
//    }
        
    if image is String {
        _imageView.path = image as? String;
    }else if image is UIImage {
        _imageView.image = image as? UIImage;
    }else{
        _imageView.image = nil;
    }
//}
    }

fileprivate func update(){

    if _imageView.image == nil {
        return;
    }
    
    var viewRect = self.bounds;
    
    let imageSize = _imageView.image?.size ?? CGSize(width: 0,height: 0);
    if imageSize.width == 0 || imageSize.height == 0 {
        return;
    }
    
    if viewRect.size.width == 0 || viewRect.size.height == 0 {
        viewRect = CGRect(x: 0, y: 0, width: imageSize.width, height: imageSize.height);
        self.frame = viewRect;
//        [self layoutIfNeeded];
    }
    var contentViewRect = viewRect;
    if imageSize.width != _preImageSize.width || imageSize.height != _preImageSize.height
        || contentViewRect.size.width != _preVieSize.width || contentViewRect.size.height != _preVieSize.height {
        _preImageSize = imageSize;
        _preVieSize = contentViewRect.size;
        var s:CGFloat = 1.0;
        if self.fill != ImageFill.fill {
            s = (_preImageSize.width/_preImageSize.height)/(contentViewRect.size.width/contentViewRect.size.height);
        }
        if (ImageFill.default == self.fill && s > 1) || self.fill == ImageFill.width {
            contentViewRect.size.width = contentViewRect.size.width;
            contentViewRect.size.height = contentViewRect.size.height / s;
            contentViewRect.origin.x = 0;
            contentViewRect.origin.y = (viewRect.size.height - contentViewRect.size.height) / 2;
        }else{
            contentViewRect.size.width = contentViewRect.size.width * s;
            contentViewRect.size.height = contentViewRect.size.height;
            contentViewRect.origin.y = 0;
            contentViewRect.origin.x = (viewRect.size.width - contentViewRect.size.width) / 2;
        }
        _imageView.frame = contentViewRect;
    
    }
}

    open override func layoutSubviews() {
       
    self.update();
    super.layoutSubviews();
}

}

